File: apps/web/src/lib/training/plan_adherence.test.ts
import { test } from 'node:test';
import assert from 'node:assert/strict';
import {
	weeklyDrift,
	missedWorkoutAdvice,
	PLAN_DRIFT_THRESHOLD,
} from './plan_adherence';

// ─────────────────────── weeklyDrift ───────────────────────

test('weeklyDrift: on-track when actual matches planned', () => {
	const d = weeklyDrift(40_000, 41_000);
	assert.equal(d.direction, 'on_track');
	assert.equal(d.flagged, false);
});

test('weeklyDrift: flags under-running past the threshold', () => {
	// 40 km planned, 28 km actual → −30% drift.
	const d = weeklyDrift(40_000, 28_000);
	assert.equal(d.direction, 'under');
	assert.equal(d.flagged, true);
	assert.ok(d.driftFraction < -PLAN_DRIFT_THRESHOLD);
});

test('weeklyDrift: flags over-running past the threshold', () => {
	// 40 km planned, 52 km actual → +30% drift (the easy-week trap).
	const d = weeklyDrift(40_000, 52_000);
	assert.equal(d.direction, 'over');
	assert.equal(d.flagged, true);
	assert.ok(d.driftFraction > PLAN_DRIFT_THRESHOLD);
});

test('weeklyDrift: just inside the threshold is not flagged', () => {
	// 40 km planned, 47 km actual → +17.5%, under the 20% bar.
	const d = weeklyDrift(40_000, 47_000);
	assert.equal(d.direction, 'on_track');
	assert.equal(d.flagged, false);
});

test('weeklyDrift: no planned volume yields a neutral, unflagged result', () => {
	const d = weeklyDrift(0, 30_000);
	assert.equal(d.direction, 'on_track');
	assert.equal(d.flagged, false);
	assert.equal(d.driftFraction, 0);
});

test('weeklyDrift: clamps negative actual to zero', () => {
	const d = weeklyDrift(40_000, -5);
	assert.equal(d.actualMetres, 0);
	assert.equal(d.direction, 'under');
});

// ─────────────────────── missedWorkoutAdvice ───────────────────────

test('missedWorkoutAdvice: base/build long run is worth making up', () => {
	const a = missedWorkoutAdvice({
		kind: 'long',
		isTaper: false,
		recoveryWeekImminent: false,
	});
	assert.equal(a.recommendation, 'make_up');
	assert.equal(a.reason, 'key_session');
});

test('missedWorkoutAdvice: skip a long run missed in the taper', () => {
	const a = missedWorkoutAdvice({
		kind: 'long',
		isTaper: true,
		recoveryWeekImminent: false,
	});
	assert.equal(a.recommendation, 'skip');
	assert.equal(a.reason, 'taper');
});

test('missedWorkoutAdvice: skip when a recovery week is imminent', () => {
	const a = missedWorkoutAdvice({
		kind: 'long',
		isTaper: false,
		recoveryWeekImminent: true,
	});
	assert.equal(a.recommendation, 'skip');
	assert.equal(a.reason, 'recovery_soon');
});

test('missedWorkoutAdvice: taper takes precedence over recovery-soon', () => {
	const a = missedWorkoutAdvice({
		kind: 'long',
		isTaper: true,
		recoveryWeekImminent: true,
	});
	assert.equal(a.reason, 'taper');
});

test('missedWorkoutAdvice: a missed quality session is just skipped', () => {
	for (const kind of ['tempo', 'interval', 'easy', 'marathon_pace']) {
		const a = missedWorkoutAdvice({ kind, isTaper: false, recoveryWeekImminent: false });
		assert.equal(a.recommendation, 'skip');
		assert.equal(a.reason, 'not_long_run');
	}
});

File: apps/mobile_android/test/plan_adherence_test.dart
import 'package:flutter_test/flutter_test.dart';
import '../lib/plan_adherence.dart';

void main() {
  group('weeklyDrift', () {
    test('on-track when actual matches planned', () {
      final d = weeklyDrift(40000, 41000);
      expect(d.direction, DriftDirection.onTrack);
      expect(d.flagged, false);
    });

    test('flags under-running past the threshold', () {
      final d = weeklyDrift(40000, 28000);
      expect(d.direction, DriftDirection.under);
      expect(d.flagged, true);
      expect(d.driftFraction < -planDriftThreshold, true);
    });

    test('flags over-running past the threshold', () {
      final d = weeklyDrift(40000, 52000);
      expect(d.direction, DriftDirection.over);
      expect(d.flagged, true);
      expect(d.driftFraction > planDriftThreshold, true);
    });

    test('just inside the threshold is not flagged', () {
      final d = weeklyDrift(40000, 47000);
      expect(d.direction, DriftDirection.onTrack);
      expect(d.flagged, false);
    });

    test('no planned volume yields a neutral, unflagged result', () {
      final d = weeklyDrift(0, 30000);
      expect(d.direction, DriftDirection.onTrack);
      expect(d.flagged, false);
      expect(d.driftFraction, 0);
    });

    test('clamps negative actual to zero', () {
      final d = weeklyDrift(40000, -5);
      expect(d.actualMetres, 0);
      expect(d.direction, DriftDirection.under);
    });
  });

  group('missedWorkoutAdvice', () {
    test('base/build long run is worth making up', () {
      final a = missedWorkoutAdvice(const MissedWorkoutInput(
          kind: 'long', isTaper: false, recoveryWeekImminent: false));
      expect(a.recommendation, MakeUpRecommendation.makeUp);
      expect(a.reason, MissedWorkoutReason.keySession);
    });

    test('skip a long run missed in the taper', () {
      final a = missedWorkoutAdvice(const MissedWorkoutInput(
          kind: 'long', isTaper: true, recoveryWeekImminent: false));
      expect(a.recommendation, MakeUpRecommendation.skip);
      expect(a.reason, MissedWorkoutReason.taper);
    });

    test('skip when a recovery week is imminent', () {
      final a = missedWorkoutAdvice(const MissedWorkoutInput(
          kind: 'long', isTaper: false, recoveryWeekImminent: true));
      expect(a.recommendation, MakeUpRecommendation.skip);
      expect(a.reason, MissedWorkoutReason.recoverySoon);
    });

    test('taper takes precedence over recovery-soon', () {
      final a = missedWorkoutAdvice(const MissedWorkoutInput(
          kind: 'long', isTaper: true, recoveryWeekImminent: true));
      expect(a.reason, MissedWorkoutReason.taper);
    });

    test('a missed quality session is just skipped', () {
      for (final kind in ['tempo', 'interval', 'easy', 'marathon_pace']) {
        final a = missedWorkoutAdvice(MissedWorkoutInput(
            kind: kind, isTaper: false, recoveryWeekImminent: false));
        expect(a.recommendation, MakeUpRecommendation.skip);
        expect(a.reason, MissedWorkoutReason.notLongRun);
      }
    });
  });
}
